### PR TITLE
Feat/add article xpression

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

**Comment by ElevenWilson** [comment]
58 words might be misspelled, please check them: amoutte, appelable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, d'etre, developpeur, DSL, d'utilisé, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, gerer, getter, GraphQL, http, JS, L'age, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, Operateur, orm, ORM, param, parametres, planete, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, résèrvé, selectionner, selectionnera, sourceName, stoquer, subtilitée, symfony, symftony, Symftony, targetName, veu, xpression, Xpression

**Comment by ElevenWilson** [comment]
45 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
45 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
45 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

File: _posts/fr/2018-07-11-xpression.md [hunk 1]
```diff
@@ -0,0 +1,473 @@
+---
+layout: post
+title: "Présentation de la librairie PHP Xpression"   
+lang: fr
+permalink: /fr/presentation-php-xpression/  
+excerpt: "En tant que développeur nous avons tous déjà eu besoin de filtrer un jeu de donnés (array, collection, API etc...). Nous allons donc découvrir la librairie Xpression qui va nous permettre de filtrer différent contenus avec une syntaxe simplifier."
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Simplifiée

File: _posts/fr/2018-07-11-xpression.md [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+authors:  
+    - amoutte  
+categories:
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Dictrine

File: _posts/fr/2018-07-11-xpression.md [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+- `non et`: 14
+- `ou`: 10
+- `ou exclusif`: 9
+- `not or`: 8
+
+Pour gérer correctement vos expressions vous pouvez utiliser les parenthèses `(` `)`.
+
+Par exemple cette expression sélectionnera les `Raccoon` ou les `Schizo` qui ont plus de 100 points.
+
+`planet='Raccoon'|name='Schizo'&point>100` et identique à `planet='Raccoon'|(name='Schizo'&point>100)`
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Est

File: _posts/fr/2018-07-11-xpression.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+ 
+Alors que l'expression suivante sélectionnera les astronautes `Raccoon` qui on plus de 100 points ou les `Schizo` qui on plus de 100 points. 
+ 
+`(planet='Raccoon'|name='Schizo')&point>100` 
+
+## Utilisation
+
+Nous allons maintenant voir dans quel cas nous pourrions utilisé cette librairie.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Utiliser

File: _posts/fr/2018-07-11-xpression.md [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+
+### Spécification
+
+Afin d'avoir une specification nous allons utiliser la classe `ClosureExpressionBuilder`.
+En effet cette class fabrique une callback qui peu être utilisé comme une spécification.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Utilisée

File: _posts/fr/2018-07-11-xpression.md [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+);
+```
+
+### Filtrer un endpoint d'API
+
+Actuellement si vous voulez filtrer votre API vous pouvez :
+ 
+ - utiliser GraphQL.
+ 
+> Ce n'est pas la solution la plus légère à implementer. N'est pas forcement adapté pour faire uniquement du filtrage de donnés.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Adaptée

File: _posts/fr/2018-07-11-xpression.md [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+ 
+> Ce n'est pas la solution la plus légère à implementer. N'est pas forcement adapté pour faire uniquement du filtrage de donnés. 
+ 
+ - récuperer les paramètres de requête manuellement et fabriquer votre query avec tous un tas de condition
+
+> la syntaxe http des paramètres n'est pas lisible et peu devenir très lourde pour des requêtes complexe.
+
+Bonne nouvelle! Si votre API utilise une des sources de donnés vu précédemment vous pouvez filtrer les donnés à l'aide d'`Xpression`.
+
+> Garder à l'esprit que Xpression remplis un rôle différent que GraphQL
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Different  de

File: _posts/fr/2018-07-11-xpression.md [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+Nous allons voir un exemple d'utilisation du [Bundle Xpression](https://github.com/Symftony/Xpression-Bundle).
+
+Installer le bundle via `composer require symftony/xpression-bundle` puis ajouter le dans symfony (AppKernel.php ou bundle.php).
+
+Il faut également activer la correction de querystring pour que les caractères réservé soit correctement utiliser.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Utilisés

File: _posts/fr/2018-07-11-xpression.md [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+```php
+<?php
+// public/index.php ou web/app.php (web/app_dev.php)
+// ...
+\Symftony\Xpression\QueryStringParser::correctServerQueryString(); // ajouter cette ligne juste avant la création de la Requete
+$request = Request::createFromGlobals();
+// ...
+``` 
+
+Pour l'utiliser il vous suffit uniquement d'ajouter l'annotation `@Xpression(expressionBuilder="odm")` au dessus du controller que vous souhaitez filtrer.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: 👏👏👏👏👏👌👌

File: _posts/fr/2018-07-11-xpression.md [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+
+Maintenant vous pouvez vous rendre sur votre URL et y ajouter votre Xpression dans `query`.
+
+```
+http://localhost/astronauts/list?query={planet\{\{Raccoons\}\}|points≥1000}
+```
+
+### Mots de la fin
+
+Je vais m'arrêter la pour la présentation de cette librairie PHP.
```
Comment (rmasclef) [inline _posts/fr/2018-07-11-xpression.md]: Là

**Comment by ElevenWilson** [comment]
46 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, endraw, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
46 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, endraw, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
46 words might be misspelled, please check them: amoutte, appellable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, endraw, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getter, GraphQL, http, JS, l'alias, lang, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, Ok, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, récuperer, refacto, sourceName, stoquer, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
45 words might be misspelled, please check them: amoutte, appelable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, endraw, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getters, GraphQL, http, JS, l'alias, lang, leakent, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, reconnait, refacto, sourceName, symfony, symftony, Symftony, targetName, xpression, Xpression

**Comment by ElevenWilson** [comment]
45 words might be misspelled, please check them: amoutte, appelable, AppKernel, ArrayCollection, ClosureExpressionBuilder, createQueryBuilder, d'API, DSL, endraw, EntityRepository, expressionBuilder, ExpressionBuilder, ExpressionBuilderAdapter, getters, GraphQL, http, JS, l'alias, lang, leakent, l'expressionBuilder, l'ORM, manyToMany, MapperExpressionBuilder, MongoDB, odm, ODM, oneToMany, orm, ORM, param, Précédence, querybuilder, queryBuilder, querystring, Raccoons, reconnait, refacto, sourceName, symfony, symftony, Symftony, targetName, xpression, Xpression

**Pull request merged**